blogposts/2021/optimizing-a-code-intelligence-commit-graph-1.md

```diff
@@ -0,0 +1,257 @@
+
+```sql
+(l.direction = 'A' AND c.commit = l.parent_commit) OR (l.direction = 'D' AND c.parent_commit = l.commit)
+```
+
+To reduce the cost of this index scan, we rewrote the recursive term within the table expression above (the second term of the `UNION`). Instead of using a `SELECT` with an `OR` condition (which had trouble using our existing set of indexes), we use `UNION` to combine the results matching each side of the `OR` clause.
+
+```sql
+SELECT c.*, l.distance + 1, l.direction
+FROM lineage l
+JOIN commits_with_lsif_data_markers c
+ON c.repository = l.repository AND l.direction = 'A' AND c."commit" = l.parent_commit
+
+UNION
+
+SELECT c.*, l.distance + 1, l.direction
+FROM lineage l
+JOIN commits_with_lsif_data_markers c
+ON c.repository = l.repository AND l.direction = 'D' AND c.parent_commit = l."commit"
+```
+
+The query plan for this new query, shown below, seems more complex at first glance. However, this query is **drastically** more efficient. The same input that required 330ms to evaluate now takes under 1ms to evaluate.
+
+<figure>
+  <img src="https://sourcegraphstatic.com/blog/commit-graph-optimizations/super-fast.png" alt="Super fast query plan"/>
+  <figcaption>Query plan of an optimized commit graph traversal visiting 100 commits.</figcaption>
+</figure>
+
+Even though we're now executing a greater number of steps, each step can be evaluated efficiently. The old, inefficient index scan has been broken into 2 different index scans: one that traverses the ancestor direction, and another that traverses the descendant direction. The simple conditionals in each query can be evaluated _without an index filter_ by each of the multicolumn indexes above. Instead of pulling back the entire commit graph on each iteration of the table expression, we pull exactly the set of rows that need to be added to the working table.
+
+This change makes the query efficient enough that we no longer have to worry about caching results, as running it on every request only contributes a negligible amount of time. Thus solving the problem [once and for all](https://www.youtube.com/watch?v=IjmtVKOAHPM)!
+
+## Lessons learned
+
+Firstly, I spoke too soon as there wouldn't be a Part 2 if we _had_ solved the problem once and for all. Secondly, sometimes the more complex the query, the more efficient it is. And lastly, the cliché holds: computers do what you tell them to. When debugging this is usually a good place to start.
```

Comment:
@efritz I'm putting words in your mouth here (did I butcher the joke??) so again feel free to edit or flesh out as you see fit 😁 It's a little abrupt as is because I'm out of my depth 😆

Reply:
```suggestion
One major take away for us was the concrete reinforcement that databases are a hugely deep subject, but are not magic. Building and maintaining an accurate mental model of how your tools work seems to be a necessary skill in software engineering, and acting with poor mental models can lead to disastrous performance results (at best).

Another major takeaway is that the cliché holds: computers do what you tell them to and nothing more. When debugging correctness and performance issues, it's good to _shed_ your previous mental model and re-assess its correctness as you dive deeper into your stack.
```